File: R/calculus.R
# Function to calculate the Jacobian using complex step differentiation
get_jacobian <- function(func, state, output_dim) {
  input_dim <- length(state)
  jacobian <- matrix(0, nrow = output_dim, ncol = input_dim)
  h <- 1e-6

  for (i in 1:input_dim) {
    state_perturbed <- as.complex(state)
    state_perturbed[i] <- state_perturbed[i] + complex(real = 0, imaginary = h)
    f_perturbed <- func(state_perturbed)
    for (j in 1:output_dim) {
      jacobian[j, i] <- Im(f_perturbed[j]) / h
    }
  }

  return(jacobian)
}

# Example function f
f <- function(x) {
  return(c(x[1]^2, x[2]^3))
}

# Example function g
g <- function(x) {
  return(c(cos(x[2]) * sin(x[1]), sin(x[1])))
}

# Test cases
test_get_jacobian <- function() {
  # Test case for function f
  x <- c(1.0, 2.0)
  jacobian <- get_jacobian(f, x, 2)
  expected <- matrix(c(2.0, 0.0, 0.0, 12.0), nrow = 2, byrow = TRUE)
  stopifnot(all.equal(jacobian, expected, tolerance = 1e-5))

  # Test case for function g
  two_pi <- 2 * pi
  theta_a_seq <- seq(0, two_pi, length.out = 100)
  theta_b_seq <- seq(0, two_pi, length.out = 100)
  
  for (theta_a in theta_a_seq) {
    for (theta_b in theta_b_seq) {
      x <- c(theta_a, theta_b)
      jacobian <- get_jacobian(g, x, 2)
      expected <- matrix(c(
        cos(x[2]) * cos(x[1]), -sin(x[2]) * sin(x[1]),
        cos(x[1]), 0.0
      ), nrow = 2, byrow = TRUE)
      stopifnot(all.equal(jacobian, expected, tolerance = 1e-5))
    }
  }
  print("All tests passed!")
}

# Run tests
test_get_jacobian()

File: haskell/calculus.hs
import Data.Complex
import Numeric.LinearAlgebra (Matrix, fromLists, toLists, (><))
import Numeric.LinearAlgebra.Data (matrix)
import Test.HUnit
import Control.Monad (forM_)

-- Function to calculate the Jacobian using complex step differentiation
getJacobian :: ([Complex Double] -> [Complex Double]) -> [Double] -> Int -> [[Double]]
getJacobian func state outputDim = toLists $ matrix outputDim $ concat jacobianElements
  where
    inputDim = length state
    h = 1e-6
    jacobianElements = [ [ (imagPart (fPerturbed !! j)) / h | j <- [0..outputDim-1] ]
                       | i <- [0..inputDim-1],
                         let statePerturbed = [ if k == i then x + (0 :+ h) else x | (x, k) <- zip (map (:+ 0) state) [0..] ],
                         let fPerturbed = func statePerturbed ]

-- Example function f
f :: [Complex Double] -> [Complex Double]
f [x1, x2] = [x1 * x1, x2 * x2 * x2]

-- Example function g
g :: [Complex Double] -> [Complex Double]
g [x1, x2] = [cos x2 * sin x1, sin x1]

-- Test case for function f
testGetJacobianF = TestCase $ do
    let x = [1.0, 2.0]
    let jacobian = getJacobian f x 2
    let expected = [[2.0, 0.0], [0.0, 12.0]]
    assertEqual "Test case f failed" expected jacobian

-- Test case for function g
testGetJacobianG = TestCase $ do
    let twoPi = 2 * pi
    forM_ [0, pi/50 .. twoPi] $ \thetaA ->
      forM_ [0, pi/50 .. twoPi] $ \thetaB -> do
        let x = [thetaA, thetaB]
        let jacobian = getJacobian g x 2
        let expected = [[cos thetaB * cos thetaA, -sin thetaB * sin thetaA],
                        [cos thetaA, 0.0]]
        assertBool ("Test case g failed for thetaA=" ++ show thetaA ++ ", thetaB=" ++ show thetaB)
                   (and $ zipWith (all . absDiff (1e-5)) jacobian expected)
  where
    absDiff tol a b = abs (a - b) < tol

main :: IO ()
main = do
  _ <- runTestTT (TestList [testGetJacobianF, testGetJacobianG])
  return ()
